File: day-4/Q2559.cpp
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> vowelStrings(vector<string>& words, vector<vector<int>>& queries) {
        vector<char> vowels = {'a', 'e', 'i', 'o', 'u'};
        unordered_set<char> vowelSet(vowels.begin(), vowels.end());
        vector<int> prefixSum(words.size() + 1, 0);
        vector<int> result;

        for (int i = 0; i < words.size(); ++i) {
            bool isValid = vowelSet.count(words[i].front()) && vowelSet.count(words[i].back());
            prefixSum[i + 1] = prefixSum[i] + (isValid ? 1 : 0);
        }

        for (auto& query : queries) {
            int l = query[0];
            int r = query[1];
            result.push_back(prefixSum[r + 1] - prefixSum[l]);
        }

        return result;
    }
};

File: day-4/Q2559.java
class Solution {
    public int[] vowelStrings(String[] words, int[][] queries) {
        int[] result = new int[queries.length];
        int[] prefix = new int[words.length + 1];

        for (int i = 0; i < words.length; ++i) {
            if (isVowel((words[i].charAt(0))) &&
                isVowel((words[i].charAt(words[i].length() - 1)))) {
                    prefix[i + 1] = prefix[i] + 1;
            } else {
                prefix[i + 1] = prefix[i];
            }
        }

        for (int i = 0; i < queries.length; ++i) {
            result[i] = prefix[queries[i][1] + 1] - prefix[queries[i][0]];
        }


        return result;
    }

    private boolean isVowel(char c) {
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
    }
}